subprojects/gradle/runner/device-provider/src/test/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClientTest.kt

```diff
@@ -0,0 +1,124 @@
+import kotlinx.coroutines.test.TestCoroutineScope
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+import java.util.LinkedList
+
+@ExperimentalCoroutinesApi
+internal class KubernetesReservationClientTest {
+
+    private val kubernetesApi = FakeKubernetesApi()
+    private val dispatcher = TestCoroutineDispatcher()
+    private fun runBlockingTest(block: suspend TestCoroutineScope.() -> Unit) {
+        dispatcher.runBlockingTest(block)
+    }
+
+    private fun client(dispatcher: CoroutineDispatcher = this.dispatcher): KubernetesReservationClient {
+        return KubernetesReservationClient(
+            androidDebugBridge = FakeAndroidDebugBridge(),
+            kubernetesApi = kubernetesApi,
+            emulatorsLogsReporter = StubEmulatorsLogsReporter,
+            loggerFactory = StubLoggerFactory,
+            reservationDeploymentFactory = FakeReservationDeploymentFactory(),
+            dispatcher = dispatcher,
+            podsQueryIntervalMs = 1L
+        )
+    }
+
+    @Test
+    fun `empty reservations - throws exception`() {
+        val client = client()
+        val exception = assertThrows<IllegalArgumentException> {
+            runBlockingTest {
+                client.claim(emptyList(), this)
+            }
+        }
+        assertThat(exception.message)
+            .isEqualTo("Must have at least one reservation but empty")
+    }
+
+    @Test
+    fun `creating deployment fail - throws exception`() {
+        val client = client()
+        val message = "Failed to create deployment"
+        kubernetesApi.createDeployment = { throw RuntimeException(message) }
+        val exception = assertThrows<RuntimeException>(message = message) {
+            runBlockingTest {
+                client.claim(listOf(ReservationData.stub()), this)
+            }
+        }
+        assertThat(exception.message)
+            .isEqualTo(message)
+    }
+
+    @Test
+    fun `claim twice - throws exception`() {
+        val client = client()
+        val exception = assertThrows<IllegalStateException> {
+            runBlockingTest {
+                // first
+                client.claim(listOf(ReservationData.stub()), this)
+
+                // second
+                client.claim(listOf(ReservationData.stub()), this)
+            }
+        }
+        assertThat(exception.message)
+            .isEqualTo("Unable to start reservation job. Already started")
+    }
+
+    @Test
+    fun `get one pod then empty - success`() {
+        val client = client()
+        val results = LinkedList(
+            listOf(
+                listOf(StubPod()),
+                emptyList()
+            ),
+        )
+
+        kubernetesApi.getPods = {
+            results.poll()
+        }
+        dispatcher.runBlockingTest {
+            client.claim(listOf(ReservationData.stub()), this)
+        }
+    }
+
+    @Test
+    fun `devices channel cancel - success`() {
```

Comment:
success case when another part of the runner did her job and cancel devices channel